ipywidgets/widgets/widget_output.py

```diff
@@ -108,23 +108,39 @@ def __enter__(self):
         """Called upon entering output widget context manager."""
         self._flush()
         ip = get_ipython()
-        if ip and hasattr(ip, 'kernel') and hasattr(ip.kernel, '_parent_header'):
+        if ip and hasattr(ip, 'kernel') and \
+               hasattr(ip.kernel, '_parent_header'):
             self.msg_id = ip.kernel._parent_header['header']['msg_id']
             self.__counter += 1
+        elif self.comm is not None and self.comm.kernel is not None and \
+               hasattr(self.comm.kernel, '_parent_header'):
+            self.msg_id = self.comm.kernel._parent_header['header']['msg_id']
+            self.__counter += 1
 
     def __exit__(self, etype, evalue, tb):
         """Called upon exiting output widget context manager."""
-        ip = get_ipython()
+        kernel = None
         if etype is not None:
+            ip = get_ipython()
             if ip:
+                kernel = ip
                 ip.showtraceback((etype, evalue, tb), tb_offset=0)
+            elif self.comm is not None and self.comm.kernel is not None:
+                kernel = self.comm.kernel
+                kernel.send_response(kernel.iopub_socket,
```

Comment:
Not sure using those kernel APIs is a good idea.

I wonder if simply re-raising the error would work? And let the kernel catch it and publish error traceback. If this doesn't work, I don't know a better solution than the current PR.

Reply:
The point behind this code is to publish the error message to the output widget (i.e., before `self.msg_id` is reset below). If we just raised the error, we wouldn't be trapping it for the output widget.

It's debatable whether trapping and showing the error in the output widget is better than raising it to the enclosing scope, but that would also be a change in behavior, and I think this PR is trying to be backwards compatible with the behavior.